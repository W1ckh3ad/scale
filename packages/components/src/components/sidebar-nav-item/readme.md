# scale-sidebar-nav-item



<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description                                                   | Type                                     | Default     |
| ----------- | ------------ | ------------------------------------------------------------- | ---------------------------------------- | ----------- |
| `bold`      | `bold`       | Bold text                                                     | `boolean`                                | `false`     |
| `condensed` | `condensed`  | Used normally for third level items, remove the bottom border | `boolean`                                | `false`     |
| `isCurrent` | `is-current` | Text gets the active color                                    | `boolean`                                | `false`     |
| `styles`    | --           | (optional) Injected jss styles                                | `StyleSheet<string \| number \| symbol>` | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
