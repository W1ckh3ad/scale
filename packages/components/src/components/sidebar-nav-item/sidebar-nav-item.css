.sidebar-nav-item {
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-bottom-color: #e0e0e0;
}

.sidebar-nav-item--active {
  position: relative;
  color: #e20074;
}
.sidebar-nav-item--active::before {
  content: '';
  display: block;
  position: absolute;
  top: 0;
  left: var(--left-current-border, 0);
  width: 0; /* The transparent border makes the 2px */
  height: 100%;
  background-color: #e20074;
  /* Accessibility: extra transparent line for Windows HCM */
  border-left: 2px solid transparent;
}

.sidebar-nav-item--condensed.sidebar-nav-item--active::before {
  left: -1px;
}

:host([nesting-level='1']) .sidebar-nav-item,
.sidebar-nav-item--bold {
  font-weight: 700;
}

::slotted(a) {
  color: currentColor;
  display: block;
  padding-top: 1rem;
  padding-right: 1rem;
  padding-bottom: 1rem;
  padding-left: var(--spacing-indent);
  border-radius: 2px;
  text-decoration: none;
  outline: none;
}

:host([nesting-level='2']) ::slotted(a) {
  padding-left: calc(var(--spacing-indent) + 0.75rem);
}

::slotted(a:hover) {
  color: #e20074;
}

::slotted(a:focus) {
  box-shadow: inset 0 0 0 2px #009de0;
}

::slotted(a:active) {
  color: #cb0068;
}

.sidebar-nav-item--condensed {
  border-bottom-width: 0;
}

.sidebar-nav-item--condensed ::slotted(a) {
  padding-top: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.5rem;
  padding-bottom: 0.5rem;
}
