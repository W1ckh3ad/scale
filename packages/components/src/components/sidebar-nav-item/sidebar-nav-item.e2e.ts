import { newE2EPage } from '@stencil/core/testing';

describe('scale-sidebar-nav-item', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<scale-sidebar-nav-item/>');
    const element = await page.find('scale-sidebar-nav-item');
    expect(element).toHaveClass('hydrated');
  });
});
