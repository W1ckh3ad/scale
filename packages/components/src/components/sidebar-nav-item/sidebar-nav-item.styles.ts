import { JssStyle } from 'jss';

import { padding } from '../sidebar-nav/sidebar-nav.styles';

export const styles: JssStyle = {
  'sidebar-nav-item': {
    borderBottomWidth: 1,
    borderBottomStyle: 'solid',
    borderBottomColor: ({ colors }) => colors.grey['300'],
    '& ::slotted(a)': {
      display: 'block',
      textDecoration: 'none',
      color: 'currentColor',
      borderRadius: 2,
      ...padding.normal,
    },
    '& ::slotted(a:hover)': {
      color: '#E20074',
    },
    '& ::slotted(a:focus)': {
      outline: 'none',
      boxShadow: 'inset 0 0 0 2px #009DE0',
    },
    '& ::slotted(a:active)': {
      color: '#CB0068',
    },
  },
  'sidebar-nav-item--current': {
    color: '#E20074',
  },
  'sidebar-nav-item--bold': {
    fontWeight: ({ type }) => type.weight_bold,
  },
  'sidebar-nav-item--condensed': {
    borderBottomWidth: 0,
    '& ::slotted(a)': {
      borderWidth: 0,
      ...padding.condensed,
    },
  },
};
