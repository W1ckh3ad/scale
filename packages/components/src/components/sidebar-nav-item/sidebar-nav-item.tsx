/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { Component, h, Prop, Host, Watch, Element } from '@stencil/core';
import classNames from 'classnames';

const SR_ACTIVE_TEXT = ' Zurzeit aktiv';

@Component({
  tag: 'scale-sidebar-nav-item',
  styleUrl: 'sidebar-nav-item.css',
  shadow: true,
})
export class SidebarNavItem {
  srOnlyElement: HTMLElement;

  @Element() el: HTMLElement;

  /** Used normally for third level items, remove the bottom border */
  @Prop() condensed: boolean = false;
  /** Bold text */
  @Prop() bold: boolean = false;
  /** Text gets the active color */
  @Prop({ mutable: true, reflect: true }) active: boolean = false;
  /**
   * Mark the child link as "current" with `aria-current=page`.
   * Provide the text hint if needed, default is: "Zurzeit aktiv"
   */
  @Prop() current: string | null = null;
  /** Nesting level within the <scale-sidebar-nav> parent, gets set automatically */
  @Prop() nestingLevel: number;
  /** (optional) Extra styles */
  @Prop() styles?: string;

  @Watch('nestingLevel')
  nestingLevelChanged(newValue: number) {
    if (newValue === 1) {
      this.bold = true;
    }
  }

  @Watch('current')
  currentChanged(newValue: string | null) {
    this.handleAriaCurrentInSlottedA(newValue);
    this.syncActiveToCurrent(newValue);
  }

  componentDidLoad() {
    this.handleAriaCurrentInSlottedA(this.current);
    if (this.current != null) {
      this.syncActiveToCurrent(this.current);
    }
  }

  /**
   * If an item is `current`, it should be `active` as well
   */
  syncActiveToCurrent(newValue: string | null) {
    this.active = newValue === null ? false : true;
  }

  /**
   * When `current` is set, this will:
   * - set the aria-current=page attribute on the link
   * - append a text-only hint for screen readers
   * so we end up with something like this:
   * <a href="..." aria-current="page">
   *    Example<span style="...visible to SR only..."> Active link</span>
   * </a>
   * @param current this.current
   */
  handleAriaCurrentInSlottedA(current: string | null) {
    const a = this.el.querySelector('a');

    if (this.srOnlyElement != null) {
      a.removeChild(this.srOnlyElement);
      this.srOnlyElement = null;
    }
    if (current === null && a != null) {
      a.removeAttribute('aria-current');
      return;
    }
    if (current != null && a != null) {
      this.srOnlyElement = this.createScreenReaderOnlySpan();
      a.appendChild(this.srOnlyElement);
      a.setAttribute('aria-current', 'page');
    }
  }

  createScreenReaderOnlySpan() {
    const text = this.current !== '' ? ` ${this.current}` : SR_ACTIVE_TEXT;
    const span = document.createElement('span');
    // .sr-only but inline
    Object.assign(span.style, {
      position: 'absolute',
      left: '-10000px',
      overflow: 'hidden',
    });
    span.textContent = text;

    return span;
  }

  render() {
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}
        <li
          part={this.getBasePartMap()}
          class={this.getCssClassMap()}
          role="listitem"
        >
          <slot />
        </li>
      </Host>
    );
  }

  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }

  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }

  getCssOrBasePartMap(mode: 'basePart' | 'css') {
    const component = 'sidebar-nav-item';
    const prefix = mode === 'basePart' ? '' : `${component}--`;

    return classNames(
      component,
      this.bold && `${prefix}bold`,
      this.condensed && `${prefix}condensed`,
      this.active && `${prefix}active`
    );
  }
}
