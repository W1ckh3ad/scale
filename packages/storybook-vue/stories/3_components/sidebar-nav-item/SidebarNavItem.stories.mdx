import { Meta, Story, Props, Preview, Description } from '@storybook/addon-docs/blocks';
import ScaleSidebarNavItem from './SidebarNavItem.vue';
import description from './sidebar-nav-item.md';

<Meta
  title='Components/Sidebar Nav Item'
  component={ScaleSidebarNavItem}
  argTypes={{
    width: {
      control: {
        type: 'radio',
        options: ['256px', '25rem', '100%']
      },
      defaultValue: '25rem',
      description: 'Example width for wrapper element'
    },
    styles: {
      control: { disable: true }
    }
  }}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleSidebarNavItem },
  props: {
    width: { type: String },
    ...ScaleSidebarNavItem.props,
  },
  computed: {
    wrapperStyles() {
      return {
        width: this.width
      }
    }
  },
  template: `
    <div :style="wrapperStyles">
      <scale-sidebar-nav>
        <scale-sidebar-nav-item
          :condensed="condensed"
          :bold="bold"
          :is-current="isCurrent"
        >
          <a href="#">Overview</a>
        </scale-sidebar-nav-item>
      </scale-sidebar-nav>
    </div>
  `
})

<Description>{description}</Description>

<Preview withSource="none">
  <Story name='Standard' args={{ bold: true }}>
    {Template.bind({})}
  </Story>
</Preview>

<Props story='Standard' />

```html
<scale-sidebar-nav aria-label="primary">
  <scale-sidebar-nav-item bold>
    <a href="#">Overview</a>
  </scale-sidebar-nav-item>
</scale-sidebar-nav>
```