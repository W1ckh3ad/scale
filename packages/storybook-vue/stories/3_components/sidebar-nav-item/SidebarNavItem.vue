<template>
  <scale-sidebar-nav-item
    :condensed="condensed"
    :bold="bold"
    :is-current="isCurrent"
    :styles="styles"
  >
    <slot></slot>
  </scale-sidebar-nav-item>
</template>

<script>
export default {
  props: {
    condensed: { type: Boolean, default: false },
    bold: { type: Boolean, default: false },
    isCurrent: { type: Boolean, default: false },
    styles: { type: Object }
  }
}
</script>