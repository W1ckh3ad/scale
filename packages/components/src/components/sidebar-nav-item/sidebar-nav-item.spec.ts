import { newSpecPage } from '@stencil/core/testing';
import { SidebarNavItem } from './sidebar-nav-item';

describe('SidebarNavItem', () => {
  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [SidebarNavItem],
      html: `<scale-sidebar-nav-item></scale-sidebar-nav-item>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
