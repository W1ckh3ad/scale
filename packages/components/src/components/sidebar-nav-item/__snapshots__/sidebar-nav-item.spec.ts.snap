// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarNavItem should match snapshot 1`] = `
<scale-sidebar-nav-item>
  <mock:shadow-root>
    <style>
      .sidebar-nav-item-0-0-1 {
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-bottom-color: #E0E0E0;
}
.sidebar-nav-item-0-0-1 ::slotted(a) {
  color: currentColor;
  display: block;
  padding-top: 1rem;
  padding-left: 2rem;
  border-radius: 2px;
  padding-right: 1rem;
  padding-bottom: 1rem;
  text-decoration: none;
}
.sidebar-nav-item-0-0-1 ::slotted(a:hover) {
  color: #E20074;
}
.sidebar-nav-item-0-0-1 ::slotted(a:focus) {
  outline: none;
  box-shadow: inset 0 0 0 2px #009DE0;
}
.sidebar-nav-item-0-0-1 ::slotted(a:active) {
  color: #CB0068;
}
.sidebar-nav-item--current-0-0-2 {
  color: #E20074;
}
.sidebar-nav-item--bold-0-0-3 {}
.sidebar-nav-item--condensed-0-0-4 {
  border-bottom-width: 0;
}
.sidebar-nav-item--condensed-0-0-4 ::slotted(a) {
  padding-top: 0.5rem;
  border-width: 0;
  padding-left: 0.75rem;
  padding-right: 0.5rem;
  padding-bottom: 0.5rem;
}
    </style>
    <li class="sidebar-nav-item-0-0-1">
      <slot></slot>
    </li>
  </mock:shadow-root>
</scale-sidebar-nav-item>
`;
